rfc/0001-crashtracker-user-visable-log.md

```diff
@@ -0,0 +1,703 @@
+It automatically collects information relevant to the characterizing and debugging the crash, including stack-traces, the crash-type (e.g. SIGSIGV, SIGBUS, etc) crash, the library version, etc.
+This RFC establishes a standardized logging format for reporting this information.
+
+### Why structured json
+As a text-based format, json can be written to standard logging endpoints.
+It is (somewhat) human readable, so users can directly interpret the crash info off their log if necessary.
+As a structured format, it avoids the ambiguity of standard semi-structured stacktrace formats (as used by e.g. Java, .Net, etc).
+Due to the use of native extensions, it is possible for a single stack-trace to include frames from multiple languages (e.g. python may call C code, which calls Rust code, etc).
+Having a single structured format allows us to work across languages.
+
+## Proposed format
+A natural language description of the proposed json format is given here.
+An example is given in Appendix A, and the schema is given in Appendix B.
+
+### Required fields
+- `errortype`:
+    Currently, the only value is "crash", but this allows for extension to capture unhandled soft-errors, e.g. "panic", "uncaught exception", etc.
+- `incomplete`:
+    Boolean `false` if the crashreport is complete (i.e. contains all intended data), `true` if there is expected missing data.
+    This can happen becasue the crashtracker is architected to stream data to an out of process receiver, allowing a partial crash report to be emitted even in the case where the crashtracker itself crashed during stack trace collection.
+    This MUST be set to `true` if any required field is missing.
+- `metadata`:
+    Metadata about the system in which the crash occurred:
+    - `library_name`
+    - `library_version`
+    - `family`
+    - `tags`:
+      A set of key:value pairs, representing any tags the crashtracking system wishes to associate with the crash.
+      Examples would include "hostname", "service", and any configuration information the system wishes to track.
+- `os_info`: 
+    The OS + processor architecture on which the crash occurred.
+- `stacktrace`: 
+    This represents the stack of the crashing thread.
+    See below for more details on how stacktraces are formatted.
+- `timestamp`:
+    The time at which the crash occurred, in ISO 8601 format.
+- `uuid`:
+    A UUID which uniquely identifies the crash.
+- `version_id`:
+    A Semver compatible ID for this format. [TODO, should it be semver?]
+
+### Optional fields
+Any field not listed as "Required" is optional.
+In order to minimize logging overhead, producers SHOULD NOT emit anything for an optional field.
+Consumers MUST accept json with elided optional fields.
+
+- `additional_stacktraces`:
+    This field contains a `Map<ThreadId, Stacktrace>`.
+    In a multi-threaded program, the collector SHOULD collect the stacktraces of all active threads, and report them here.
```

Comment:
There's value in not directly tying it to ThreadId only. Like, you might want to print the managed _and_ the unmanaged stacktrace (i.e. the C-stack trace and the actual stacktrace within the runtime).

Reply:
For .NET I merged the managed and native stacktraces (sorting the frames by stackpointer) but I don't know if that makes sense for all languages

Reply:
That makes only sense if both are tied together, which isn't the case in e.g. PHP, which has two separate stacks for PHP and C, with no really discernible connection.

Reply:
Not all stack traces are the same.
Instead of 1 format handling all - lets make sure we accommodate all languages - current and future.

e.g. we can version a stacktrace schema.

Reply:
Modified as per the comment below.